lib/processors/pr0nprocessor.py

```diff
@@ -0,0 +1,264 @@
+        fname = "data/pr0n/backend/{0[0]}/{0[1]}/{0}.pkl".format(img_hash)
+        with open(fname, 'rb') as fd:
+            return pickle.load(fd)
+
+    def create_engine(self):
+        rbp = RandomBinaryProjections('rbp', 8)
+        return Engine(128, lshashes=[rbp])
+
+    def read_pr0n(self):
+        num_added = 0
+        for dirname, _, filenames in os.walk('./data/pr0n/backend/'):
+            if filenames:
+                for filename in filenames:
+                    if not filename.endswith('pkl'):
+                        continue
+                    filepath = os.path.join(dirname, filename)
+                    filehash = filename.split('.')[0]
+                    with open(filepath, 'rb') as fd:
+                        data = pickle.load(fd)
+                        if data['pose'] != 0:
+                            continue
+                        self.pr0n_engine.store_vector(data['face_hash'],
+                                                      filehash)
+                        num_added += 1
+        self.logger.info("Added backend pictures: {}".format(num_added))
+        if not num_added:
+            self.logger.error("PRON: No pictures added... try running "
+                              "./scripts/create_pr0n_database.py")
+
+    def filter_photos(self, photos):
+        for photo in photos:
+            if not photo.get('faces'):
+                continue
+            elif not any(f.get('pose') == 0 for f in photo['faces']):
+                continue
+            elif not any(f.get('face_hash', None) is not None
+                         for f in photo['faces']):
+                continue
+            yield photo
+
+    @gen.coroutine
+    def process(self, user_data):
+        """
+        Go through all the tagged facebook pictures from the user and creates a
+        nearpy structure for easy similarity lookup
+        """
+        self.logger.info("Processing user: {}".format(user_data.userid))
+        if not user_data.data.get('fbphotos'):
+            return False
+        photos_me_and_friends = IT.chain.from_iterable(
+            user_data.data['fbphotos'].values()
+        )
+        # NOTE: is there a better way to figure out the name of the current
+        # user within the facebook tagging universe?  i assume one of the other
+        # facebook scrapers is going to extract that information somewhere and
+        # we'll be able to use it here instead of doing this crazy tag counting
+        me_candidates = Counter(t['name']
+                                for photo in user_data.data['fbphotos']['me']
+                                for face in photo['faces']
+                                for t in face['tags'])
+        me, _ = me_candidates.most_common(1)[0]
+
+        photos = list(self.filter_photos(photos_me_and_friends))
+        if not photos:
+            return False
+        names_to_scores = {}
+        images_to_scores = {}
+        user_engine = self.create_engine()
```

Comment:
there can be more than just one engine.  it creates a `nearpy` object to do approximate nearest neighbor searches (ie: http://nearpy.io/)